File: tst/size_pct.c
#include "pico.h"
#include "../check.h"

int main (void) {
    pico_init(1);
    pico_set_window_title("Size - Fullscreen");

    Pico_Rel_Rect r = { '%', {0.5, 0.5, 0.5, 0.5}, PICO_ANCHOR_C, NULL };

    // phy: normal -> double -> half -> normal
    puts("WINDOW");
    {
        puts("normal");
        Pico_Abs_Dim cur = pico_get_window_dim();
        Pico_Rel_Dim dim = { '!', {cur.w * 1, cur.h * 1}, NULL };
        pico_set_window_dim(&dim);
        pico_output_clear();
        pico_output_draw_rect(&r);
        _pico_check("size_pct-01");
    }
    {
        puts("double");
        Pico_Abs_Dim cur = pico_get_window_dim();
        Pico_Rel_Dim dim = { '!', {cur.w * 2, cur.h * 2}, NULL };
        pico_set_window_dim(&dim);
        pico_output_clear();
        pico_output_draw_rect(&r);
        _pico_check("size_pct-02");
    }
    {
        puts("half");
        Pico_Abs_Dim cur = pico_get_window_dim();
        Pico_Rel_Dim dim = { '!', {cur.w * 0.25, cur.h * 0.25}, NULL };
        pico_set_window_dim(&dim);
        pico_output_clear();
        pico_output_draw_rect(&r);
        _pico_check("size_pct-03");
    }
    {
        puts("normal");
        Pico_Abs_Dim cur = pico_get_window_dim();
        Pico_Rel_Dim dim = { '!', {cur.w * 2, cur.h * 2}, NULL };
        pico_set_window_dim(&dim);
        pico_output_clear();
        pico_output_draw_rect(&r);
        _pico_check("size_pct-04");
    }

    // phy: normal -> full -> normal
    {
        puts("full");
        pico_set_window_fs(1);
        pico_output_clear();
        pico_output_draw_rect(&r);
        //_pico_check("size_pct-05");
    }
    {
        puts("normal");
        pico_set_window_fs(0);
        pico_output_clear();
        pico_output_draw_rect(&r);
        _pico_check("size_pct-06");
    }

    // log: normal -> double -> half -> normal
    puts("WORLD");
    {
        puts("normal");
        Pico_Abs_Dim cur;
        cur = pico_get_view_dim(NULL);
        Pico_Rel_Dim dim = { '!', {cur.w * 1, cur.h * 1}, NULL };
        pico_set_view_dim(NULL, &dim);
        pico_output_clear();
        pico_output_draw_rect(&r);
        _pico_check("size_pct-07");
    }
    {
        puts("double");
        Pico_Abs_Dim cur;
        cur = pico_get_view_dim(NULL);
        Pico_Rel_Dim dim = { '!', {cur.w * 2, cur.h * 2}, NULL };
        pico_set_view_dim(NULL, &dim);
        pico_output_clear();
        pico_output_draw_rect(&r);
        _pico_check("size_pct-08");
    }
    {
        puts("half");
        Pico_Abs_Dim cur;
        cur = pico_get_view_dim(NULL);
        Pico_Rel_Dim dim = { '!', {cur.w * 0.25, cur.h * 0.25}, NULL };
        pico_set_view_dim(NULL, &dim);
        pico_output_clear();
        pico_output_draw_rect(&r);
        _pico_check("size_pct-09");
    }
    {
        puts("normal");
        Pico_Abs_Dim cur;
        cur = pico_get_view_dim(NULL);
        Pico_Rel_Dim dim = { '!', {cur.w * 2, cur.h * 2}, NULL };
        pico_set_view_dim(NULL, &dim);
        pico_output_clear();
        pico_output_draw_rect(&r);
        _pico_check("size_pct-10");
    }

    pico_init(0);
    return 0;
}

File: lua/tst/size_pct.lua
require 'pico.check'

pico.init(true)
pico.set.window { title="Size - Fullscreen" }

local r = {'%', x=0.5, y=0.5, w=0.5, h=0.5}

-- phy: normal -> double -> half -> normal
print("WINDOW")
do
    print("normal")
    pico.set.layer("window")
    local win = pico.get.scene().dim
    pico.set.scene { dim={'!', w=win.w*1, h=win.h*1} }
    pico.set.layer("world")
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("size_pct-01")
end

do
    print("double")
    pico.set.layer("window")
    local win = pico.get.scene().dim
    pico.set.scene { dim={'!', w=win.w*2, h=win.h*2} }
    pico.set.layer("world")
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("size_pct-02")
end

do
    print("half")
    pico.set.layer("window")
    local win = pico.get.scene().dim
    pico.set.scene { dim={'!', w=win.w*0.25, h=win.h*0.25} }
    pico.set.layer("world")
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("size_pct-03")
end

do
    print("normal")
    pico.set.layer("window")
    local win = pico.get.scene().dim
    pico.set.scene { dim={'!', w=win.w*2, h=win.h*2} }
    pico.set.layer("world")
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("size_pct-04")
end

-- phy: normal -> full -> normal
do
    print("full")
    pico.set.window { fullscreen=true }
    pico.output.clear()
    pico.output.draw.rect(r)
    -- pico.check("size_pct-05")
end

do
    print("normal")
    pico.set.window { fullscreen=false }
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("size_pct-06")
end

-- log: normal -> double -> half -> normal
print("WORLD")
do
    print("normal")
    local view = pico.get.scene()
    pico.set.scene { dim={'!', w=view.dim.w*1, h=view.dim.h*1} }
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("size_pct-07")
end

do
    print("double")
    local view = pico.get.scene()
    pico.set.scene { dim={'!', w=view.dim.w*2, h=view.dim.h*2} }
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("size_pct-08")
end

do
    print("half")
    local view = pico.get.scene()
    pico.set.scene { dim={'!', w=view.dim.w*0.25, h=view.dim.h*0.25} }
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("size_pct-09")
end

do
    print("normal")
    local view = pico.get.scene()
    pico.set.scene { dim={'!', w=view.dim.w*2, h=view.dim.h*2} }
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("size_pct-10")
end

pico.init(false)
